### PR TITLE
Support Image() by digest

### LINKED ISSUE #34: _parse_image_url() doesn't support image digests
A recent deploy of a Hive operator failed with:

`Image does not exist: quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e`

The image exists:

```
docker pull quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e
quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e: Pulling from ****/pagerduty-operator-registry
Digest: sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e
Status: Image is up to date for quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e
quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e
```

```
13:58:43 [2021-04-16 17:58:43] [ERROR] [saasherder.py:_check_image:543] - [saas-pagerduty-operator/pagerduty-operator] https://github.com/openshift/pagerduty-operator/blob/master/hack/olm-registry/olm-artifacts-template.yaml: Image does not exist: quay.io/****/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e
13:58:43 Traceback (most recent call last):
13:58:43   File "/usr/local/bin/qontract-reconcile", line 33, in <module>
13:58:43     sys.exit(load_entry_point('reconcile==0.2.2', 'console_scripts', 'qontract-reconcile')())
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 829, in __call__
13:58:43     return self.main(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 782, in main
13:58:43     rv = self.invoke(ctx)
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1259, in invoke
13:58:43     return _process_result(sub_ctx.command.invoke(sub_ctx))
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 1066, in invoke
13:58:43     return ctx.invoke(self.callback, **ctx.params)
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/core.py", line 610, in invoke
13:58:43     return callback(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/environ.py", line 15, in f_environ
13:58:43     f(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/environ.py", line 15, in f_environ
13:58:43     f(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/binary.py", line 18, in f_binary
13:58:43     f(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/binary.py", line 60, in f_binary_version
13:58:43     f(*args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/click-7.1.2-py3.6.egg/click/decorators.py", line 21, in new_func
13:58:43     return f(get_current_context(), *args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/cli.py", line 758, in openshift_saas_deploy
13:58:43     saas_file_name, env_name, gitlab_project_id)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/cli.py", line 399, in run_integration
13:58:43     func_container.run(dry_run, *args, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/defer.py", line 15, in func_wrapper
13:58:43     return func(*args, defer=defer, **kwargs)
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/openshift_saas_deploy.py", line 69, in run
13:58:43     if not saasherder.validate_promotions():
13:58:43   File "/usr/local/lib/python3.6/site-packages/reconcile-0.2.2-py3.6.egg/reconcile/utils/saasherder.py", line 928, in validate_promotions
13:58:43     commit_sha = item['commit_sha']
13:58:43 TypeError: 'NoneType' object is not subscriptable
```

SREP haven't hit this issue yet because on cluster operators are deployed via selector sync set inside the OC template. Saasherder doesn't validate the CatalogSource images inside the selector sync set [here](https://github.com/app-sre/qontract-reconcile/blob/master/reconcile/utils/saasherder.py#L524). This is different for Hive operators as we deploy the CatalogSource directly in the OC template.

It looks like that method is the issue.

Parsing a digest returns the following structure:

```
{'scheme': 'docker://', 'registry': 'docker.io', 'port_colon': None, 'port': None, 'registry_slash': None, 'repository': 'library', 'repo_slash': None, 'image': 'sha256', 'tag_colon': ':', 'tag': 'bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e'}
```

cc/ @2uasimojo @BumbleFeng @apahim

### PR DESCRIPTION
**Pull request by 2uasimojo**
Support initializing an Image with a URI that is specified "by digest", e.g. `quay.io/app-sre/pagerduty-operator-registry@sha256:bc1ed82a75f2ca160225b8281c50b7074e7678c2a1f61b1fb298e545b455925e`.

Of note, when using such an Image:
- The `digest` property works without going to the network.
- The `manifest` property works BUT at least for quay.io, the output uses schemaVersion 2, whereas it is 1 when using a by-tag URI.
- The `url_tag` property raises a (new exported type of) Exception called NoTagForImageByDigest since there may be more than one tag for a given image.
- You're still allowed to provide a tag override, in which case `url_tag` will work. But you're responsible for making sure the tag and the manifest are for the same image!
- The `__str__` method returns a by-digest URI (having properly defaulted things like registry etc. if necessary).

Closes #34

### CODE REVIEW

File: sretoolbox/container/image.py [hunk 1]
```diff
@@ -47,6 +54,11 @@ def __init__(self, url, tag_override=None, username=None, password=None,
         else:
             self.tag = tag_override
 
+        self._cache_digest = None
+        # If the URL was by-digest, we can cache this right away.
+        if image_data['digest']:
+            self._cache_digest = f"sha256:{image_data['digest']}"
```
Comment (apahim) [inline sretoolbox/container/image.py:60]: the `sha256:` should be coming from the `image_data['digest']`... more on that ahead

File: sretoolbox/container/image.py [hunk 2]
```diff
@@ -218,6 +233,8 @@ def _parse_image_url(image_url):
             r'(?P<repository>[\w\-]+)?'  # Repository (optional)
             r'(?(repository)(?P<repo_slash>/))'  # Slash, if repo is present
             r'(?P<image>[\w\-.]+)'  # Image path (mandatory)
+            r'(?P<by_digest>@sha256:)?'  # Present iff it's a by-digest URI
+            r'(?(by_digest)(?P<digest>[0-9a-f]{64}))'  # Digest SHA: 64 lowercase hex chars
```
Comment (apahim) [inline sretoolbox/container/image.py:237]: Awesome, thanks for that!
I'd reccomend using the `@` as the marker and add the `sha256:` to the `<digest>` group. Then use the `<digest>` group to conditionally-negativelly look for the `tag_colon` ahead.

My suggestion here is:

```diff
-            r'(?P<by_digest>@sha256:)?'  # Present iff it's a by-digest URI
-            r'(?(by_digest)(?P<digest>[0-9a-f]{64}))'  # Digest SHA: 64 lowercase hex chars
-            r'(?P<tag_colon>:)?'  # Tag colon (optional)
-            r'(?(tag_colon)(?P<tag>[\w\-.]+))'  # Tag (if tag colon is present)
+            # Digest marker "@" (optional)
+            r'(?P<digest_at>@)?'
+            # Digest (if @ is present)
+            r'(?(digest_at)(?P<digest>sha256:[0-9a-f]{64}))'
+            # Tag colon (if digest not is present, optional)
+            r'(?(digest)|(?P<tag_colon>:))?'
+            # Tag (if tag colon is present)
+            r'(?(tag_colon)(?P<tag>[\w\-.]+))'
```

File: sretoolbox/container/image.py [hunk 3]
```diff
@@ -243,7 +260,8 @@ def _parse_image_url(image_url):
             else:
                 image_url_struct['repository'] = None
 
-        if image_url_struct.get('tag') is None:
+        # By-digest URIs don't use tags; but otherwise default to `latest` if absent
+        if image_url_struct.get('tag') is None and not image_url_struct.get('by_digest'):
```
Comment (apahim) [inline sretoolbox/container/image.py:264]: `... and image_url_struct.get('by_digest') is None`, for consistency & correctness

**Review by apahim — CHANGES_REQUESTED** [review]
Hey Eric, thank you for the contribution! Awesome work here, I agree with the expected behaviour.
Besides a few formatting issues making the CI unhappy, I leaft a couple of comments, please let me know what do you think.

File: sretoolbox/container/__init__.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 """
 
 from sretoolbox.container.image import Image
+from sretoolbox.container.image import NoTagForImageByDigest
```
[the remark this answers is not in the document]
Comment (2uasimojo) [reply sretoolbox/container/__init__.py:6]: I’ll try to figure out the import from the test — I was having trouble doing that without the export. 

However, IMHO having the exception be part of the interface makes sense to me...

File: sretoolbox/container/image.py (hunk 1 above)
Comment (2uasimojo) [reply sretoolbox/container/image.py:60]: Ack, I’m on board.

File: sretoolbox/container/image.py (hunk 2 above)
Comment (2uasimojo) [reply sretoolbox/container/image.py:237]: Dig, will do. I’ll add a test to prove you can’t put a tag after a digest as well, which I think you could probably do as written.

File: sretoolbox/container/image.py (hunk 3 above)
Comment (2uasimojo) [reply sretoolbox/container/image.py:264]: Roger that

File: sretoolbox/container/image.py (hunk 2 above)
Comment (2uasimojo) [reply sretoolbox/container/image.py:237]: > I’ll add a test to prove you can’t put a tag after a digest as well, which I think you could probably do as written.

Well, this isn't strict enough for such a test to make sense, and this isn't the PR to start adding such strictness. Maybe later :)

File: sretoolbox/container/__init__.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (2uasimojo) [reply sretoolbox/container/__init__.py:6]: I couldn't figure out how to do this import, so I hacked around it. LMK if there's a better way.

[the remark this answers is not in the document]
Comment (apahim) [reply sretoolbox/container/__init__.py]: Fair enough, let's keep it.

File: sretoolbox/container/image.py (hunk 2 above)
Comment (apahim) [reply sretoolbox/container/image.py]: Sounds good

**Review by apahim — APPROVED** [review]

**Comment by apahim** [comment]
Lgtm, thank you!

**Comment by 2uasimojo** [comment]
Thanks @apahim!

Can we get this promoted through qontract-reconcile/app-interface ASAP please?

**Pull request merged**